funtofem/mphys/mphys_meld.py

```diff
@@ -269,89 +316,206 @@ def compute_jacvec_product(self, inputs, d_inputs, d_outputs, mode):
-                    # df_s/du_s^T * psi = - dL/du_s^T * psi
-                    self.meld.applydLduSTrans(d_out, d_in)
+                    if "x_aero0" in d_inputs:
+                        # df_s/dx_a0^T * psi = - psi^T * dL/dx_a0 in F2F terminology
+                        prod = np.zeros(
+                            len(body.aero_coord_indices), dtype=TransferScheme.dtype
+                        )
+                        body.meld.applydLdxA0(d_out, prod)
+                        d_inputs["x_aero0"][body.aero_coord_indices] -= np.array(
+                            prod, dtype=float
+                        )
 
-                    for i in range(3):
-                        d_inputs["u_struct"][i :: self.struct_ndof] -= np.array(
-                            d_in[i::3], dtype=float
+                    if "x_struct0" in d_inputs:
+                        # df_s/dx_s0^T * psi = - psi^T * dL/dx_s0 in F2F terminology
+                        prod = np.zeros(
+                            len(body.struct_coord_indices), dtype=TransferScheme.dtype
+                        )
+                        body.meld.applydLdxS0(d_out, prod)
+                        d_inputs["x_struct0"][body.struct_coord_indices] -= np.array(
+                            prod, dtype=float
                         )
 
-                if "f_aero" in d_inputs:
-                    # df_s/df_a^T psi = - dL/df_a^T * psi = -dD/du_s * psi
-                    prod = np.zeros(self.aero_nnodes * 3, dtype=TransferScheme.dtype)
-                    self.meld.applydDduS(d_out, prod)
-                    d_inputs["f_aero"] -= np.array(prod, dtype=float)
 
-                if "x_aero0" in d_inputs:
-                    # df_s/dx_a0^T * psi = - psi^T * dL/dx_a0 in F2F terminology
-                    prod = np.zeros(self.aero_nnodes * 3, dtype=TransferScheme.dtype)
-                    self.meld.applydLdxA0(d_out, prod)
-                    d_inputs["x_aero0"] -= np.array(prod, dtype=float)
+class MeldBodyInstance:
+    """
+    Class that helps split OpenMDAO input/output indices for multiple bodies,
+    with a separate MELD instance for each
+    """
+
+    def __init__(
+        self,
+        comm,
+        isym=-1,
+        n=200,
+        beta=0.5,
+        aero_gid=None,  # list of aero grid IDs to couple
+        aero_nnodes=None,
+        struct_gid=None,  # list of struct grid IDs to couple
+        struct_nnodes=None,
+        struct_ndof=None,
+        linearized=False,
+    ):
+        # determine input/output indices for the current body
+        if (
+            struct_gid is not None
+        ):  # use grid ids from struct_builder's get_tagged_indices
+            self.struct_coord_indices = []  # for coordinates
+            self.struct_dof_indices = []  # for dof
+            for i in range(3):  # accounts for xyz of each grid coordinate
+                self.struct_coord_indices = np.hstack(
+                    [self.struct_coord_indices, struct_gid * 3 + i]
```

Comment:
This line is breaking for me when `struct_gid` is a `list` rather than an `array`

Reply:
You should probably make sure this is cast to an array first with `np.atleast1d` in case the user passes back the indices in a list from `get_tagged_indices`

Reply:
I can't push to your fork @Asthelen but I think this code would fix the issue:

```python
            # account for xyz of each node (create array of [x_0, ..., x_N, y_0, ..., y_N, z_0, ..., z_N])
            struct_gid = np.atleast_1d(struct_gid)
            num_struct_gid = len(struct_gid)
            self.struct_coord_indices = np.tile(struct_gid, 3)
            for ii in range(3):
                self.struct_coord_indices[
                    ii * num_struct_gid : (ii + 1) * num_struct_gid
                ] += ii

            # account for each structural DOF of each node (create array of [u_0, ..., u_N, v_0, ..., v_N, w_0, ..., w_N, ...])
            self.struct_dof_indices = np.zeros(num_struct_gid * struct_ndof)
            for ii in range(struct_ndof):
                self.struct_dof_indices[
                    ii * num_struct_gid : (ii + 1) * num_struct_gid
                ] = (struct_gid * struct_ndof + ii)

            self.struct_coord_indices = list(
                np.sort(self.struct_coord_indices).astype("int")
            )
            self.struct_dof_indices = list(
                np.sort(self.struct_dof_indices).astype("int")
            )
```

Reply:
I had to tweak the first part to make it consistent with what was there... thoughts on this instead?

```
# account for xyz of each node (i.e., [x_0, y_0, z_0, ..., x_N, y_N, z_N])
struct_coord_indices = np.zeros(struct_nnodes * 3, dtype=int)                    
for ii in range(3):                                                              
    struct_coord_indices[                                                        
        ii * struct_nnodes : (ii + 1) * struct_nnodes                            
    ] = struct_gid * 3 + ii                                                      
                                                                                 
# account for each structural DOF of each node (i.e., [u_0, v_0, w_0, ..., u_N, v_N, w_N])
struct_dof_indices = np.zeros(struct_nnodes * struct_ndof, dtype=int)            
for ii in range(struct_ndof):                                                    
    struct_dof_indices[                                                          
        ii * struct_nnodes : (ii + 1) * struct_nnodes                            
    ] = (struct_gid * struct_ndof + ii)                                          
                                                                                 
struct_coord_indices = list(                                                     
    np.sort(struct_coord_indices)                                                
)                                                                                
struct_dof_indices = list(                                                       
    np.sort(struct_dof_indices)                                                  
)
```

Then something similar for aero indices too. I also have the np.atleast_1d around the get_tagged_indices calls now. The np.atleast_1d fixed the list vs. array issue, but maybe this is better than using hstack.

Reply:
Sorry--I should've tested that first. struct_nnodes is all nodes whereas your

`num_struct_gid = len(struct_gid)`

Is only the number of structural grid nodes on the current body. I'm not sure if there'd be other issues with it.

With the np.atleast_1d fix, I'd be okay with leaving that section of code the same and just adding the comments (e.g., pointing to [x_0, y_0, z_0, ..., x_N, y_N, z_N])?

Reply:
I think I might be misunderstanding the ordering, if you have 3 nodes, a,b,c, in the body then `struct_gid = [a, b, c]`, won't this method give you `self.struct_coord_indices = [a, b, c, a+1, b+1, c+1, a+2, b+2, c+2]`? So your array contains the index of all the x coordinates first, then all the y coordinates, then the z?

Reply:
struct_coord_indices would be the array indices in x_struct0 that correspond to that body. So since x_struct0 is flattened as [x0, y0, z0, x1, y1, z1, x2, y2, z2, ...], if a certain body contains nodes 0 and 2 (i.e., struct_gid = [0,2]), then struct_coord_indices would be [0, 1, 2, 6, 7, 8].

Reply:
I don't think the current `hstack` method will give you that ordering though, e.g:
```python
import numpy as np
struct_gid = [0, 2]
struct_gid=np.at_least1d(struct_gid)
struct_coord_indices = []
for i in range(3):
    struct_coord_indices = np.hstack([struct_coord_indices, struct_gid*3+i])
print(struct_coord_indices)
```
Gives `[0. 6. 1. 7. 2. 8.]`

Reply:
I don't think the current `hstack` method will give you that ordering though, e.g:
```python
import numpy as np
struct_gid = [0, 2]
struct_gid=np.at_least1d(struct_gid)
struct_coord_indices = []
for i in range(3):
    struct_coord_indices = np.hstack([struct_coord_indices, struct_gid*3+i])
print(struct_coord_indices)
```
Gives `[0. 6. 1. 7. 2. 8.]`

Reply:
Yeah, this sort a bit later turns it into [0, 1, 2, 6, 7, 8] though:

```
            self.struct_coord_indices = list(
                np.sort(self.struct_coord_indices).astype("int")
            )
```

Reply:
I agree, @A-CGray's method seems like a more straightforward alternative here. Plus it gets rid of the unnecessary sort step at the end.

I think that the last line should actually be `struct_coord_indices[ii::3] = 3 * struct_gid + ii`

Reply:
Yeah I agree too... I figured it could've been cleaned up, but didn't pull the trigger on that. Latest push has that change, and I also changed *_gid to *_node_ids.